src/XPurchase/VirtoCommerce.XPurchase/Services/CartProductService.cs

```diff
@@ -32,49 +35,71 @@ IItemService productService
             _mapper = mapper;
         }
 
-        public async Task<IEnumerable<CartProduct>> GetCartProductsByIdsAsync(CartAggregate cartAggr, string[] ids, string additionalResponseGroups = null)
+        public async Task<IEnumerable<CartProduct>> GetCartProductsByIdsAsync(CartAggregate cartAggr, string[] ids)
         {
-            if (cartAggr == null)
-            {
-                throw new ArgumentNullException(nameof(cartAggr));
-            }
-            if (ids == null)
-            {
-                throw new ArgumentNullException(nameof(ids));
+            if (cartAggr == null)
+            {
+                throw new ArgumentNullException(nameof(cartAggr));
+            }
+            if (ids == null)
+            {
+                throw new ArgumentNullException(nameof(ids));
             }
 
-            var defaultResponseGroups = ItemResponseGroup.ItemAssets | ItemResponseGroup.ItemInfo | ItemResponseGroup.Outlines | ItemResponseGroup.Seo;
-
-            var result = new List<CartProduct>();
-            var products = await _productService.GetByIdsAsync(ids, (defaultResponseGroups | EnumUtility.SafeParseFlags(additionalResponseGroups, ItemResponseGroup.None)).ToString());
-            if (!products.IsNullOrEmpty())
-            {
-                var loadInventoriesTask = _inventorySearchService.SearchInventoriesAsync(new InventorySearchCriteria
-                {
-                    ProductIds = ids,
-                    //Do not use int.MaxValue use only 10 items per requested product
-                    //TODO: Replace to pagination load
-                    Take = Math.Min(ids.Length * 10, 500)
-                });
-
-                var pricesEvalContext = _mapper.Map<PriceEvaluationContext>(cartAggr);
+            var products = await GetProductsByIdsAsync(ids);
+            var result = await AddPricesAndInventorysToCartProductAsync(cartAggr, products.ToArray());
+            return result;
+        }
+
+        protected virtual async Task<IEnumerable<CatalogProduct>> GetProductsByIdsAsync(string[] ids)
+        {
+            return await _productService.GetByIdsAsync(ids, defaultResponseGroups.ToString());
+        }
+
+        protected virtual async Task<IEnumerable<CartProduct>> AddPricesAndInventorysToCartProductAsync(CartAggregate cartAggr, CatalogProduct[] products)
+        {
+            var pageNumber = 1;
+            var skip = 0;
+            var take = pageSize;
+
+            var result = new List<CartProduct>();
+            var allLoadInventories = new List<InventoryInfo>();
+
+            if (!products.IsNullOrEmpty())
+            {
+                var ids = products.Select(x => x.Id).ToArray();

```

Comment:
a bit cleaner paging code can be lifted from
https://github.com/VirtoCommerce/vc-module-experience-api/blob/dev/src/VirtoCommerce.ExperienceApiModule.DigitalCatalog/Middlewares/EvalProductsInventoryMiddleware.cs